### ISSUE TITLE
Products with Qty Uses Decimal and Min Quantity < 1 broken.

### ISSUE THREAD

**Issue by markalston** [magento/magento2#5570]
Trying to add products to cart which have "Qty Uses Decimals" = Yes and  "Minimum Qty Allowed in Shopping Cart" < 1  with quantity less than one from the products page fails.  They can however be added with quantity one and then edited down to the desired quantity on the /checkout/cart page.

The issue seems to be due to the data-validate settings in the input field.

On the product page and the mini cart ends up like: 

```
<div class="control">
<input id="qty" class="input-text qty" type="number" data-validate="{"required-number":true,"validate-item-quantity":{"minAllowed":1}}" title="Qty" value="0.01" maxlength="12" name="qty">
</div>
```

But on /checkout/cart where I am able to edit the quantities to below 1 it looks like this:

```
<div class="control qty">
<input id="cart-91-qty" class="input-text qty" type="number" data-role="cart-item-qty" data-validate="{required:true,'validate-greater-than-zero':true}" maxlength="12" title="Qty" size="4" value="0.2" data-cart-item-id="avangard-2" name="cart[91][qty]">
</div>
```

**Comment by antboiko** [comment]
Hi @markalston , thanks for reporting this issue. We've created internal ticket MAGETWO-55238.

Best,
Anton.

**Comment by jamematen** [comment]
any news on this?

**Comment by markalston** [comment]
Looking for any update on this issue. This really breaks our site as we sell about a hundred items in fractional amounts.

**Comment by sajidunnar** [comment]
Hello

any status ?

**Comment by ashburystudios** [comment]
Hi all, Any update on this. Need this ASAP to work on a new site and cant move forward. Thank you, Mike

**Comment by markalston** [comment]
I think the problem is this line in ProductView.php but I am not sure.  This line seems to disallow min quantities less than 1.

magento/module-catalog-inventory/Block/Plugin/ProductView.php:        $params['minAllowed']  = max((float)$stockItem->getQtyMinAllowed(), 1);

**Comment by markalston** [comment]
The following changes to afterGetQuanittyValidators in vendor/magento/module-catalog-inventory/Block/Plugin/ProductView.php fixes the problem on the item page and the edit cart page.  It is broken in the mini cart which seems to be completely ignoring the quantity decimal option.

```
public function afterGetQuantityValidators(
        \Magento\Catalog\Block\Product\View $block,
        array $validators
    ) {
        $stockItem = $this->stockRegistry->getStockItem(
            $block->getProduct()->getId(),
            $block->getProduct()->getStore()->getWebsiteId()
        );

        $params = [];
        $params['minAllowed']  = max((float)$stockItem->getQtyMinAllowed(), 0);
        if ($stockItem->getQtyMaxAllowed()) {
            $params['maxAllowed'] = $stockItem->getQtyMaxAllowed();
        }
        if ($stockItem->getEnableQtyIncrements() && $stockItem->getQtyIncrements() > 0) {
            $params['qtyIncrements'] = (float)$stockItem->getQtyIncrements();
        }
        $validators['validate-item-quantity'] = $params;

        return $validators;
    }

```

**Comment by markalston** [comment]
Found a fix for the minicart problem which turns out to be a separate bug really where the minicart will only allow integer quantities. This is due to the HTML5 input type="number" in magento/module-checkout/view/frontend/web/template/minicart/item/default.html.

By adding step="any" after type="number" it allows decimal values in the minicart.

**NOTE for those using porto or other themes. You will probably have to fix this in your themes override file instead of the base file.

```
<div class="details-qty qty">
                    <label class="label" data-bind="i18n: 'Qty', attr: {
                           for: 'cart-item-'+item_id+'-qty'}"></label>
                    <input data-bind="attr: {
                           id: 'cart-item-'+item_id+'-qty',
                           'data-cart-item': item_id,
                           'data-item-qty': qty,
                           'data-cart-item-id': product_sku
                           }, value: qty"
                           type="number"
                           step="any"
                           size="4"
                           class="item-qty cart-item-qty"
                           maxlength="12"/>
                    <button data-bind="attr: {
                           id: 'update-cart-item-'+item_id,
                           'data-cart-item': item_id,
                           title: $t('Update')
                           }"
                            class="update-cart-item"
                            style="display: none">
                        <span data-bind="i18n: 'Update'"></span>
                    </button>
                </div>
```

**Comment by markalston** [comment]
One more file needs an edit.  In module-checkout/Controller/Sidebar/UpdateItemQty.php the quantity is cast to an integer which breaks all decimal items.  Changing that cast to double (combined with the fix above) fixes the minicart for decimal items.  This is really an entirely different bug than the one I originally submitted so I can open a new issue with all these fixes if needed.

**Comment by markalston** [comment]
Found another place with decimal issues.  In module-catalog/templates/product/view/addtocart.phtml the qty input has type="number" it needs to have step="any" added (at least for decimal items).  Since this is only client side validation I just added step="any" under the type="number" and it fixes the problem.  

It would probably be best to add a check for decimal item being set and then add step="any" here and in magento/module-checkout/view/frontend/web/template/minicart/item/default.html in that case only. 

Actually, ideally something should probably be done like the following wherever there is a type="number" input for item quantity (note, I haven't tested this code yet) :

```
<?php if ($stockItem->getEnableQtyIncrements() && $stockItem->getQtyIncrements() > 0): ?> 
           step="<?php (float)$stockItem->getQtyIncrements() ?>"
<?php elseif ($stockitem->getIsQtyDecimal() ): ?>
           step="any";
<?php else: ?>
           step="1";
<?php endif; ?>

```

looks like the changes would also have to be made in the following files as well as each of these uses type="number" on a quantity input field:
module-checkout/view/frontend/templates/cart/item/default.phtml
module-checkout/view/frontend/templates/cart/item/configure/updatecart.phtml
module-wishlist/view/frontend/templates/item/column/cart.phtml

**Comment by alena-marchenko** [comment]
Hi @markalston 

Fix for MAGETWO-56943 is merged to develop branch, closing the issue.
Thank you.

**Comment by Ctucker9233** [comment]
@alena-marchenko Which version of Magento will have MAGETWO-56943 included?

**Comment by JohnNDK** [comment]
/vendor/magento/module-checkout/view/frontend/web/template/minicart/content.html

Summery count in minicart needs to be updated as well.
Knockout only validates 1 and 0, not 0.x, which results in only the amount being displayed with no text.

Not sure how to validate a float with Knockout.

```<div class="items-total">
            <span class="count"><!-- ko text: getCartParam('summary_count') --><!-- /ko --></span>
            <!-- ko if: getCartParam('summary_count') == 1 -->
                <!-- ko i18n: 'item' --><!-- /ko -->
            <!-- /ko -->
            <!-- ko if: getCartParam('summary_count') > 1 -->
                <!-- ko i18n: 'items' --><!-- /ko -->
            <!-- /ko -->
        </div>```

**Comment by markalston** [comment]
These fixes don't seem to solve the problem.  While the test suite might work, without adding the step="any" lines as noted in my earlier comment, end users will not be allowed to type decimal numbers into the qty boxes.  

The following files still need step="any" added to work:

module-catalog/templates/product/view/addtocart.phtml
module-checkout/view/frontend/web/template/minicart/item/default.html
module-checkout/view/frontend/templates/cart/item/default.phtml
module-checkout/view/frontend/templates/cart/item/configure/updatecart.phtml
module-wishlist/view/frontend/templates/item/column/cart.phtml

**Comment by Ctucker9233** [comment]
@magento-engcom-team Any progress on this?

**Comment by markalston** [comment]
They closed the issue even though they haven't fixed it.  My fixes above work fine for me.  Hopefully they will reopen this.  If not, we will probably have to submit a new issue referencing this one.

**Comment by Hexmage** [comment]
The bug appears to only have been fixed for Magento 2.2.

**Comment by aabumuslimov** [comment]
Find the solution in https://github.com/magento/magento2ce/commit/f0655b0bde5065b0ade3df9f01178b549fc20999
